src/go/MONIT/alert.go

```diff
@@ -0,0 +1,556 @@
+func tabulate() {
+
+	w := new(tabwriter.Writer)
+	w.Init(os.Stdout, 8, 8, 0, '\t', 0)
+	defer w.Flush()
+
+	fmt.Fprintf(w, "\n %s\t\t%s\t\t%s\t\t%s\t\t%s\t\t%s\t\t%s\n", "NAME", "SERVICE", "TAG", "SEVERITY", "STARTS", "ENDS", "DURATION")
+
+	for _, each := range allAlertData {
+		if filteredAlerts[each.Name] == 1 {
+			fmt.Fprintf(w, " %s\t\t%s\t\t%s\t\t%s\t\t%s",
+				each.Name,
+				each.Service,
+				each.Tag,
+				each.Severity,
+				timeDiff(time.Now(), each.StartsAt, 0),
+			)
+			if each.EndsAt == time.Unix(maxtstmp, 0).UTC() {
+				fmt.Fprintf(w, "\t\t%s", "Undefined")
+				fmt.Fprintf(w, "\t\t%s\n", "Undefined")
+			} else {
+				fmt.Fprintf(w, "\t\t%s", timeDiff(time.Now(), each.EndsAt, 0))
+				fmt.Fprintf(w, "\t\t%s\n", timeDiff(each.StartsAt, each.EndsAt, 1))
+			}
+
+		}
+	}
+}
+
+//Function for printing alert's details in JSON format
+func jsonPrintDetails() {
+
+	b, err := json.Marshal(alertDetails[name])
+
+	if err != nil {
+		log.Printf("Unable to convert Detailed JSON Data, error: %v\n", err)
+		return
+	}
+
+	fmt.Println(string(b))
+}
+
+//Function to get all keys of type map[string]interface{}
+func getkeys(m map[string]interface{}) []string {
+
+	keys := make([]string, 0, len(m))
+	for key := range m {
+		keys = append(keys, key)
+	}
+	return keys
+}
+
+//Function for printing alert's details in Plain text format
+func detailPrint() {
+
+	if alert, ok := alertDetails[name]; ok {
+		labelsKeys := getkeys(alertDetails[name].Labels)
+		sort.Strings(labelsKeys)
+
+		for _, each := range labelsKeys {
+			switch each {
+			case "alertname":
+				fmt.Printf("NAME: %s\n", alertDetails[name].Labels[each])
+				fmt.Printf("LABELS\n")
+			case "service", "tag", "severity":
+				fmt.Printf("\t%s: %s\n", each, alertDetails[name].Labels[each])
+			}
+		}
+
+		fmt.Printf("ANNOTATIONS\n")
+		for key, value := range alert.Annotations {
+			fmt.Printf("\t%s: %s\n", key, value)
+		}
+	} else {
+		fmt.Printf("%s alert not found\n", name)
+	}
+
+}
+
+//Function running all logics
+func run() {
+
+	var amdata amData
+	get(&amdata)
+	mergeData(amdata)
+	sortAlert()
+	filter()
+
+	if *jsonOutput {
+		if *details {
+			jsonPrintDetails()
+		} else {
+			jsonPrint()
+		}
+	} else {
+		if *details {
+			detailPrint()
+		} else {
+			tabulate()
+		}
+	}
+}
+
+func parseConfig() {
+	//Default Severity Levels in case no config file is provided
+	if severityConfig == "" {
+		sLevel.SeverityLevels = make(map[string]int)
+		sLevel.SeverityLevels["info"] = 0
+		sLevel.SeverityLevels["warning"] = 1
+		sLevel.SeverityLevels["medium"] = 2
+	} else {
+		jsonFile, e := os.Open(severityConfig)
+		if e != nil {
+			fmt.Println("Severity Config File not found, error:", e)
+		}
+		defer jsonFile.Close()
+		decoder := json.NewDecoder(jsonFile)
+		err := decoder.Decode(&sLevel)
+		if err != nil {
+			fmt.Println("Severity Config JSON File can't be loaded, error:", err)
+		}
+	}
+}
+
+func main() {
+
+	flag.StringVar(&cmsmonURL, "url", "", "CMS Monit URL")
```

Comment:
Rahul, I wrote already that all input arguments can be now part of config. Please adjust that, i.e. move input arguments to config and then adjust parseConfig to get proper defaults, if necessary.

Reply:
Valentin, 
Sorry, I have made the required changes. The config file looks like this 
```
{           
    "cmsmonURL" : "https://cms-monitoring.cern.ch",
    "names": ["NAMES", "LABELS", "ANNOTATIONS"],
    "columns": ["NAME", "SERVICE", "TAG", "SEVERITY", "STARTS", "ENDS", "DURATION"],
    "verbose" : 0,
    "severity" : {
            "info" : 0,
            "warning" : 1,
            "medium" : 2,
            "high" : 3,
            "critical" : 4,
            "urgent" : 5
        }
}
```
1) CONFIG_PATH Environment Variable is required.
2) I have made our alert CLI Tool's columns configurable as well.
3) Proper defaults are set in case no Config file is provided or some values are not present in the config file. Defaults are:- 
```
//Defaults in case no config file is provided
configJSON.CMSMONURL = "https://cms-monitoring.cern.ch"
configJSON.Names = []string{"NAMES", "LABELS", "ANNOTATIONS"}
configJSON.Columns = []string{"NAME", "SERVICE", "TAG", "SEVERITY", "STARTS", "ENDS", "DURATION"}
configJSON.Verbose = 0
configJSON.SeverityLevels = make(map[string]int)
configJSON.SeverityLevels["info"] = 0
configJSON.SeverityLevels["warning"] = 1
configJSON.SeverityLevels["medium"] = 2
```